snuba_sdk/query_visitors.py

```diff
@@ -396,12 +397,23 @@ def _combine(self, query: "main.Query", returns: Mapping[str, None]) -> None:
             raise InvalidQuery("query must have at least one expression in select")
 
         # Non-aggregate expressions must be in the groupby if there is an aggregate
-        non_aggregates = []
+        non_aggregates: List[Union[Column, CurriedFunction, Function]] = []
         has_aggregates = False
+
+        # Legacy queries use aliases as Columns, so we need to compile all the possible
```

Comment:
Not sure I get what use case this refers to. 
For `use aliases as Columns` do you mean that we would find identifier in the select clause of the legacy query that are transformed into `Column` by the legacy to snql translator?

If yes, what do we need the additional condition `exp.alias:` for at line 409 ?

Reply:
The best example is the test case I added. Some legacy queries have expressions in the select clause and the aggregation clause, e.g. `{"selected_columns": ["x"], "aggregations": [["fn", ..., "x"]]...`. There is no groupby necessary in this query since there is only the aggregation selected, but the SDK would incorrectly think that `"x"` was a raw column and complain that it wasn't in the groupby clause.

Reply:
Still I don't get your condition.
Line 410 adds a column to `alias_columns` if the expression is a function, it is aggregate and it has an alias. 
Isn't this something that could be checked in the same loop at line 412?
Or is the select looking like this after the legacy translator resolved the aggregation?
```
["first_session_started", "last_session_started", "min(started)", ""max(started)" ]
```
If this is the case, could you please comment with an explanation of this use case. Reading the test case does not show what the reason for this behavior is.

Also isn't this a general problem we can encounter in validation potentially anywhere by having aliases in the expressions ?

Reply:
> Or is the select looking like this after the legacy translator resolved the aggregation?

Yes that's what the select looks like. Then the validation fails because it thinks `"first_session_started"` and `"last_session_started"` should be in the groupby.

In order to check that, I need to know the aliases of all the aggregates so in the second loop I can see if a given select expression is an alias of an aggregate or a column that should be in the groupby.

As for a general problem, the answer is "yes, but it hackily works right now". It only becomes an issue with these validators because they need to understand the difference between a Column and "something that looks like a Column but is actually an alias". Nowhere else in the code does that distinction matter (right now). I've added a tech debt task to better handle these situations and make this more explicit in the SDK, but that is not something I intend to tackle in this PR.

Reply:
Ok, could you please add a comment explaining the corner case we are dealing with. 

Also for the general case, can we just get rid of aliases entirely once the legacy translator is gone ?

Reply:
That's a discussion point. That was my original intention when I started the SDK but I'm starting to wonder if we should support them for usability reasons.